saw-remote-api/src/SAWServer/JVMCrucibleSetup.hs

```diff
@@ -0,0 +1,197 @@
+    go (SetupAlloc name ty False Nothing) =
+      error $ "cannot allocate type: " ++ show ty
+    go (SetupPointsTo src tgt) = get >>= \env -> lift $
+      do ptr <- getSetupVal env src
+         tgt' <- getSetupVal env tgt
+         error "nyi: points-to"
+    go (SetupExecuteFunction args) =
+      get >>= \env ->
+      lift $ traverse (getSetupVal env) args >>= jvm_execute_func bic opts
+    go (SetupPrecond p) = get >>= \env -> lift $
+      getTypedTerm env p >>= jvm_precond
+    go (SetupPostcond p) = get >>= \env -> lift $
+      getTypedTerm env p >>= jvm_postcond
+
+    opts = defaultOptions
+    save name val = modify' (\(env, cenv) -> (Map.insert name val env, cenv))
+
+    getSetupVal ::
+      (Map ServerName ServerSetupVal, CryptolEnv) ->
+      CrucibleSetupVal (P.Expr P.PName) ->
+      JVMSetupM (MS.SetupValue CJ.JVM)
+    getSetupVal _ NullValue = JVMSetupM $ return $ MS.SetupNull ()
+                              {-
+    getSetupVal env (ArrayValue elts) =
+      do elts' <- mapM (getSetupVal env) elts
+         JVMSetupM $ return $ MS.SetupArray () elts'
+    getSetupVal env (FieldLValue base fld) =
+      do base' <- getSetupVal env base
+         JVMSetupM $ return $ MS.SetupField () base' fld
+    getSetupVal env (ElementLValue base idx) =
+      do base' <- getSetupVal env base
+         JVMSetupM $ return $ MS.SetupElem () base' idx
+    getSetupVal _ (GlobalInitializer name) =
+      JVMSetupM $ return $ MS.SetupGlobalInitializer () name
+    getSetupVal _ (GlobalLValue name) =
+      JVMSetupM $ return $ MS.SetupGlobal () name
+         -}
+    getSetupVal (env, _) (ServerValue n) = JVMSetupM $
+      resolve env n >>=
+      \case
+        Val x -> return x -- TODO add cases for the server values that
+                          -- are not coming from the setup monad
+                          -- (e.g. surrounding context)
+    getSetupVal (_, cenv) (CryptolExpr expr) = JVMSetupM $
+      do res <- liftIO $ getTypedTermOfCExp fileReader (biSharedContext bic) cenv expr
+         -- TODO: add warnings (snd res)
+         case fst res of
+           Right (t, _) -> return (MS.SetupTerm t)
+           Left err -> error $ "Cryptol error: " ++ show err -- TODO: report properly
+    getSetupVal _ _sv = error $ "unrecognized setup value" -- ++ show sv
+
+    getTypedTerm ::
+      (Map ServerName ServerSetupVal, CryptolEnv) ->
+      P.Expr P.PName ->
+      JVMSetupM TypedTerm
+    getTypedTerm (_, cenv) expr = JVMSetupM $
+      do res <- liftIO $ getTypedTermOfCExp fileReader (biSharedContext bic) cenv expr
+         -- TODO: add warnings (snd res)
+         case fst res of
+           Right (t, _) -> return t
+           Left err -> error $ "Cryptol error: " ++ show err -- TODO: report properly
+
+    resolve env name =
+       case Map.lookup name env of
+         Just v -> return v
+         Nothing -> error "Server value not found - impossible!" -- rule out elsewhere
```

Comment:
Argo is now using the standard Galois Panic bit - perhaps this should call that?

Reply:
I think so, though it'll require some refactoring. I created #85 to track that.